beacon_node/eth2-libp2p/src/rpc/handler.rs

```diff
@@ -269,9 +273,26 @@ where
         &mut self.listen_protocol
     }
 
+    /// Initiates the handler's shutdown.
+    ///
+    /// Drops all requests queued for dialing.  After the shutdown has begun, inbound requests are
+    /// rejected but all established streams are driven to completion.
+    pub fn shutdown(&mut self) {
+        if self.shutting_down && !self.dial_queue.is_empty() {
+            warn!(self.log, "RPC handler shutting down"; "unsent_requests" => self.dial_queue.len());
+            // drop requests queued for dialing
+            self.dial_queue.clear();
```

Comment:
Its very important (especially for sync) that we tell the application (Lighthouse) if/when a request fails. 

So if we kill any request, we have to make sure we send a response. Perhaps a new RPCError which is "HandlerShutdown" or something. 

I imagine we could add a function to the handler like `shutdown_request(req: Request)` which adds an error message to `self.pending_errors` which could get consumed in the handler's poll().

Reply:
I imagine the KeepAlive has some time before actually shutting down, so we should have time to drain these errors from the `pending_errors` queue.

Reply:
I just reviewed your other PR. We probably want to keep the dial queue RPC's. The reason being, is that when we go into shutdown, we might queue the Goodbye message. We still want to send this, so I think we should process the dial_queue messages. If we are going to do that, then we should also keep the `outbound_upgrade` streams. 

I think this will be fine as long we prevent any new rpc via the `send_event()` function as you've done here